python/cudf/cudf/core/dataframe.py

```diff
@@ -437,21 +437,13 @@ def __getitem__(self, arg):
             return frame._slice(row_spec.key)
         elif isinstance(row_spec, indexing_utils.ScalarIndexer):
             result = frame._gather(row_spec.key, keep_index=True)
-            # Attempt to turn into series.
-            try:
-                # Behaviour difference from pandas, which will merrily
-                # turn any heterogeneous set of columns into a series if
-                # you only ask for one row.
-                new_name = result.index[0]
-                result = Series._concat(
-                    [result[name] for name in column_names],
-                    index=result.keys(),
-                )
-                result.name = new_name
-                return result
-            except TypeError:
-                # Couldn't find a common type, just return a 1xN dataframe.
-                return result
+            new_name = result.index[0]
+            result = Series._concat(
+                [result[name] for name in column_names],
+                index=result.keys(),
+            )
+            result.name = new_name
+            return result
```

Comment:
Do you want to gate this behind pandas-compat? Note also that this is a breaking change in that some cudf-only workflows will now no-longer work.

Additionally, should the `TypeError` here report a useful error message (recommending to use a singleton index/slice to get a 1xN dataframe, for example)?

Reply:
> Do you want to gate this behind pandas-compat? Note also that this is a breaking change in that some cudf-only workflows will now no-longer work.

Do we have to support this since we are anyway not able to match pandas here? After going through https://github.com/rapidsai/cudf/issues/11477, I feel we should just raise an useful error message rather that gating this inconsistency with pandas-compat/

> Additionally, should the `TypeError` here report a useful error message (recommending to use a singleton index/slice to get a 1xN dataframe, for example)?

+1 to raising this useful error message.


cc: @bdice | @shwina any thoughts?

Reply:
cc: @beckernick | @bdice | @shwina any thoughts?

Reply:
I’m not entirely sure what to say, but I think I would propose this:

- Make breaking changes (like raising errors) only for pandas compat mode
- If desired, raise deprecation warnings in “normal” cudf mode and then change behavior later